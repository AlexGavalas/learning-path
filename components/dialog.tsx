import { FC, useRef, useEffect } from 'react';
import { AnimatePresence, motion, HTMLMotionProps } from 'framer-motion';

import { useOnClickOutside } from '../hooks/use-on-click-outside';

type DialogSize = 'md';

interface DialogProps extends HTMLMotionProps<'dialog'> {
    onClickOutside?: () => void;
    size?: DialogSize;
}

const variants = {
    hidden: {
        opacity: 0,
    },
    enter: {
        opacity: 1,
    },
};

const FOCUSABLE_ELEMENTS =
    'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])';

const DialogContent: FC<DialogProps> = ({
    onClickOutside = () => void 0,
    size,
    children,
    ...props
}) => {
    const modalRef = useRef<HTMLDialogElement>(null);

    useOnClickOutside(modalRef, onClickOutside);

    // Hook to return focus to the trigger element
    useEffect(() => {
        if (!modalRef.current || !props.open) return;

        const previousFocusedElement = document.activeElement as HTMLElement;

        return () => {
            // process.nextTick is used to focus the element
            // after framer-motion finishes the exit animation
            process.nextTick(() => {
                previousFocusedElement?.focus();
            });
        };
    }, [props.open]);

    useEffect(() => {
        if (!modalRef.current) return;

        // Disable body scroll
        document.body.style.overflow = 'hidden';

        const focusableContent =
            modalRef.current.querySelectorAll(FOCUSABLE_ELEMENTS);

        const firstFocusableEl = focusableContent[0];

        const lastFocusableEl = focusableContent[focusableContent.length - 1];

        const handleKeyPress = (e: KeyboardEvent) => {
            if (e.key === 'Escape') {
                return onClickOutside();
            }

            const isTabPressed = e.key === 'Tab' || e.keyCode === 9;

            if (!isTabPressed) return;

            if (e.shiftKey) {
                // if shift key pressed for shift + tab combination
                if (document.activeElement === firstFocusableEl) {
                    (lastFocusableEl as HTMLElement).focus();
                    e.preventDefault();
                }
            } else if (document.activeElement === lastFocusableEl) {
                // if tab key is pressed

                (firstFocusableEl as HTMLElement).focus();
                e.preventDefault();
            }
        };

        document.addEventListener('keydown', handleKeyPress);

        (firstFocusableEl as HTMLElement).focus();

        return () => {
            document.removeEventListener('keydown', handleKeyPress);

            // Reset body scroll
            document.body.style.overflow = 'auto';
        };
    }, [props.open, onClickOutside]);

    return (
        <motion.dialog
            ref={modalRef}
            initial="hidden"
            animate="enter"
            exit="hidden"
            variants={variants}
            transition={{ type: 'linear' }}
            className={`fixed inset-0 rounded bg-black ${
                size === 'md' ? 'w-[75%]' : ''
            }`}
            {...props}
        >
            {children}
        </motion.dialog>
    );
};

export const Dialog: FC<DialogProps> = (props) => {
    return (
        <AnimatePresence>
            {props.open && <DialogContent {...props} />}
        </AnimatePresence>
    );
};
