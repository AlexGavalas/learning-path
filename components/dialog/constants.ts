import { Variants } from 'framer-motion';

export const variants: Variants = {
    hidden: {
        opacity: 0,
    },
    enter: {
        opacity: 1,
    },
};
