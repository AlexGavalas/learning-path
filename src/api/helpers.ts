import axios from 'axios';

export const fetchFileFromStorage = async (
    filename: string,
): Promise<string> => {
    const fileServerUrl = String(process.env.PUBLIC_FILE_SERVER_URL);

    const response = await axios.get(`${fileServerUrl}/${filename}`, {
        responseType: 'text',
    });

    return response.data as string;
};
