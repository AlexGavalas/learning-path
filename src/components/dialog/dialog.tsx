import { FC, useRef, useEffect } from 'react';
import { createPortal } from 'react-dom';
import { AnimatePresence, motion, HTMLMotionProps } from 'framer-motion';

import { useOnClickOutside } from '~hooks/use-on-click-outside';
import { variants } from './constants';

type DialogSize = 'md';

interface DialogProps extends HTMLMotionProps<'dialog'> {
    onClickOutside?: (event: MouseEvent | TouchEvent | KeyboardEvent) => void;
    size?: DialogSize;
}

const FOCUSABLE_ELEMENTS =
    'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])';

const DialogContent: FC<DialogProps> = ({
    onClickOutside = () => void 0,
    size,
    children,
    ...props
}) => {
    const modalRef = useRef<HTMLDialogElement | null>(null);
    const previousFocusedElementRef = useRef(
        document.activeElement as HTMLElement | null,
    );

    useOnClickOutside(modalRef, onClickOutside);

    useEffect(() => {
        const handler = (e: Event) => e.preventDefault();

        document.addEventListener('scroll', handler, { passive: false });
        document.addEventListener('wheel', handler, { passive: false });

        return () => {
            document.removeEventListener('scroll', handler);
            document.removeEventListener('wheel', handler);
        };
    }, []);

    useEffect(() => {
        if (!props.open) return;

        const rootDiv = document.body.firstChild as HTMLElement;

        // Disable interactions with the rest of the app
        rootDiv.inert = true;

        const focusableContent =
            modalRef.current?.querySelectorAll(FOCUSABLE_ELEMENTS) ?? [];

        const firstFocusableEl = focusableContent[0];

        const lastFocusableEl = focusableContent[focusableContent.length - 1];

        const handleKeyPress = (e: KeyboardEvent) => {
            if (e.key === 'Escape') {
                return onClickOutside(e);
            }

            const isTabPressed = e.key === 'Tab';

            if (!isTabPressed) return;

            if (e.shiftKey) {
                // if shift key pressed for shift + tab combination
                if (document.activeElement === firstFocusableEl) {
                    (lastFocusableEl as HTMLElement).focus();
                    e.preventDefault();
                }
            } else if (document.activeElement === lastFocusableEl) {
                // if tab key is pressed
                (firstFocusableEl as HTMLElement).focus();
                e.preventDefault();
            }
        };

        document.addEventListener('keydown', handleKeyPress);

        (firstFocusableEl as HTMLElement)?.focus();

        const previousFocusedElement = previousFocusedElementRef.current;

        return () => {
            document.removeEventListener('keydown', handleKeyPress);

            // Restore app interactivity
            rootDiv.inert = false;

            // Return focus to the trigger element
            // process.nextTick is used to focus the element
            // after framer-motion finishes the exit animation
            process.nextTick(() => {
                previousFocusedElement?.focus();
            });
        };
    }, [props.open, onClickOutside]);

    return createPortal(
        <motion.dialog
            ref={modalRef}
            initial="hidden"
            animate="enter"
            exit="hidden"
            variants={variants}
            transition={{ type: 'linear' }}
            open={props.open}
            className={`fixed inset-0 rounded bg-black ${
                size === 'md' ? 'w-[75%]' : ''
            }`}
        >
            {children}
        </motion.dialog>,
        document.body,
    );
};

export const Dialog: FC<DialogProps> = (props) => (
    <AnimatePresence>
        {props.open && <DialogContent {...props} />}
    </AnimatePresence>
);
