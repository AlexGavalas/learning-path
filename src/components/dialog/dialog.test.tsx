import { axe } from 'jest-axe';

import { renderWithUser, screen } from '~test/helpers';
import { Dialog } from './dialog';

describe('<Dialog />', () => {
    describe('when it is closed', () => {
        it('renders', () => {
            renderWithUser(<Dialog>content</Dialog>);

            expect(document.body).toMatchSnapshot();
        });
    });

    describe('when it is open', () => {
        it('renders', () => {
            renderWithUser(<Dialog open>content</Dialog>);

            expect(document.body).toMatchSnapshot();
        });

        it('is accessible', async () => {
            renderWithUser(<Dialog open>content</Dialog>);

            const a11yResults = await axe(document.body);

            expect(a11yResults).toHaveNoViolations();
        });

        it('can be navigated using tab in a circular way', async () => {
            const { user } = renderWithUser(
                <Dialog open>
                    <input id="1" />
                    <input id="2" autoFocus />
                    <input id="3" />
                </Dialog>,
            );

            expect(document.activeElement?.id).toBe('2');

            await user.tab();

            expect(document.activeElement?.id).toBe('3');

            await user.tab();

            expect(document.activeElement?.id).toBe('1');
        });

        it('can be navigated backwards using shift+tab', async () => {
            const { user } = renderWithUser(
                <Dialog open>
                    <input id="1" autoFocus />
                    <input id="2" />
                    <input id="3" />
                </Dialog>,
            );

            expect(document.activeElement?.id).toBe('1');

            await user.keyboard('{Shift>}');
            await user.tab();

            expect(document.activeElement?.id).toBe('3');
        });

        describe('when user presses Esc', () => {
            it('calls onClickoutside', async () => {
                const mockOnClickoutside = jest.fn();

                const { user } = renderWithUser(
                    <Dialog onClickOutside={mockOnClickoutside} open>
                        content
                    </Dialog>,
                );

                await user.keyboard('{Escape}');

                expect(mockOnClickoutside).toHaveBeenCalledTimes(1);
            });
        });

        describe('when user clicks outside', () => {
            it('calls onClickoutside', async () => {
                const mockOnClickoutside = jest.fn();

                const { user } = renderWithUser(
                    <Dialog onClickOutside={mockOnClickoutside} open>
                        content
                    </Dialog>,
                );

                await user.click(screen.getByRole('dialog'));

                expect(mockOnClickoutside).toHaveBeenCalledTimes(0);

                await user.click(document.body);

                expect(mockOnClickoutside).toHaveBeenCalledTimes(1);
            });
        });
    });
});
