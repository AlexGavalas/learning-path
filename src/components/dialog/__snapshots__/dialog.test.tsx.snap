// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dialog /> when it is closed renders 1`] = `
<body>
  <div />
</body>
`;

exports[`<Dialog /> when it is open renders 1`] = `
<body>
  <div />
  <dialog
    class="fixed inset-0 rounded bg-black "
    open=""
    style="opacity: 0;"
  >
    content
  </dialog>
</body>
`;
